### PR TITLE
POC: Generate Icon components from SVGs

### LINKED ISSUE #81682: Create a POC for the new icons package

### PR DESCRIPTION
**Pull request by Clarity-89**
This PR repurposes this repo for the new icons package, with a POC demonstrating how this package might function. 

**How it works**
1. SVGs for the icons are uploaded to the `svg` folder.
2. `yarn icons:create` converts the SVGs into TypeScript React components inside the `src/icons` folder. Each component is wrapped in an `IconBase` component for easier customisation of the icons. A custom template in `templates/icon.js` is used to generate each component.
3. `yarn build` compiles the generated code and creates esm + cjs bundles in the `dist` folder. This is mostly a copy/paste of the same logic from `@grafana` packages.
4. [Not implemented] Once the package is published, the new icons become available. 

**How to test locally, within Grafana**

Due to issues with package linking in yarn version 3 and above, additional steps are required to enable it to work correctly:
1. Clone this repo and check out the `svg-to-component` branch.
2. Run `yarn install && yarn build`.
3. Inside the cloned repo, run `yarn link`.
4. Navigate to the Grafana folder.
5. Run `yarn link -p <relative-path-to-the-icons-package>`. In the package.json it'd create a new resolution: ```"resolutions": {
    "@grafana/icons": "portal:/absolute-path-to-the-package"
  }```
6. Add the package to dependencies and run `yarn install`: 
```
"devDependencies": {
  "@grafana/icons": "*",
}
```
7. Import the icons inside Grafana: `import { AngleRight } from "@grafana/icons";`. The new icon API is largely similar to the one of the current `Icon` component. 

**TODOs**

- [ ] CI for publishing the package and its new versions. The publish should happen on the package.json version increment. 
- [ ] A Playground to be able to preview the icons (or should it be done in the Sage website instead?)

**Questions**
- Do all the icons needs to be wrapped in `forwardRef`? We currently do it for the `Icon` component, but I wonder if it's used? It will add quite a bit of complexity to the icon template. 
- Do we need both cjs and esm builds? 

Fixes: https://github.com/grafana/grafana/issues/81682.

### CODE REVIEW

File: rollup.config.ts [hunk 1]
```diff
@@ -0,0 +1,38 @@
+import resolve from "@rollup/plugin-node-resolve";
```
Comment (Clarity-89) [inline rollup.config.ts:1]: This file is copied from Grafana, so some parts might be not necessary.

File: src/IconBase.tsx [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+}: IconProps) => {
+  const svgSize = getSvgSize(size);
+
+  return (
+    <svg
+      xmlns="http://www.w3.org/2000/svg"
+      viewBox="0 0 24 24"
+      height={svgSize}
+      width={svgSize}
+      style={{
```
Comment (Clarity-89) [inline src/IconBase.tsx:46]: It seemed an overkill to add emotion just for this style declaration, so I made it inline. Style customisation can still be done via the `className` prop and these inline styles are not meant to be changed in any case.

File: rollup.config.ts [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+import path from "path";
+import dts from "rollup-plugin-dts";
+import esbuild from "rollup-plugin-esbuild";
+import externals from "rollup-plugin-node-externals";
+import pkg from "./package.json" assert { type: "json" };
```
Comment (Clarity-89) [inline rollup.config.ts:6]: This `assert` declaration was necessary to get rollup to run.

File: package.json [hunk 4]
```diff
@@ -1,8 +1,50 @@
...
   "version": "1.0.0",
   "description": "Icons for Grafana",
   "repository": "https://github.com/grafana/icons.git",
   "author": "Grafana Labs",
-  "license": "Apache-2.0"
+  "license": "Apache-2.0",
+  "scripts": {
+    "icons:clean": "rm -rf ./src/icons",
+    "icons:create": "npx @svgr/cli ./svg",
+    "build": "tsc -p ./tsconfig.build.json && rollup -c rollup.config.ts --configPlugin typescript"
```
Comment (academo) [inline package.json:11]: Do you need to run `tsc` here since rollup already handles the typescript compilation?

File: src/IconBase.tsx [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+  size?: IconSize;
+  /** Render the title element with the provided text.
+   * More info: https://developer.mozilla.org/en-US/docs/Web/SVG/Element/title
+   */
+  title?: string;
+  /** Color of the icon. Defaults to "currentColor" */
+  color?: string;
+}
+
+function getSvgSize(size: IconSize) {
```
Comment (academo) [inline src/IconBase.tsx:17]: There will be cases (I'm 100% sure) where we need custom sizes.

File: package.json [hunk 6]
```diff
@@ -1,8 +1,50 @@
...
+  "name": "@grafana/icons",
   "version": "1.0.0",
   "description": "Icons for Grafana",
   "repository": "https://github.com/grafana/icons.git",
   "author": "Grafana Labs",
-  "license": "Apache-2.0"
+  "license": "Apache-2.0",
+  "scripts": {
+    "icons:clean": "rm -rf ./src/icons",
+    "icons:create": "npx @svgr/cli ./svg",
```
Comment (academo) [inline package.json:10]: do you know if svgr sanitizes SVG for XSS at conversion? if not, can we make sure we do that with some static tool (so we avoid having to add some runtime dependency to the icons)

File: package.json [hunk 7]
```diff
@@ -1,8 +1,50 @@
...
+    "esbuild": "^0.20.0",
+    "prettier": "^3.2.4",
+    "rollup": "^4.9.6",
+    "rollup-plugin-dts": "^6.1.0",
+    "rollup-plugin-esbuild": "^6.1.1",
+    "rollup-plugin-node-externals": "^7.0.1",
+    "ts-node": "^10.9.2",
+    "typescript": "^5.3.3"
+  },
+  "dependencies": {
```
Comment (academo) [inline package.json:46]: these probably need to be peerDependencies. For example now grafana uses React 18. We can use something like this https://github.com/grafana/grafana/blob/main/packages/grafana-data/package.json#L96

File: src/IconBase.tsx [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+      height={svgSize}
+      width={svgSize}
+      style={{
+        display: "inline-block",
+        fill: color,
+        flexShrink: 0,
+        lineHeight: 0,
+        verticalAlign: "middle",
+      }}
+      className={className}
```
Comment (academo) [inline src/IconBase.tsx:53]: you can remove `className` and it'll take it from  `...props` here

File: src/index.ts [hunk 9]
```diff
@@ -0,0 +1,2 @@
+export { type IconProps } from "./IconBase";
+export * from "./icons/index";
```
Comment (academo) [inline src/index.ts:2]: I prefer we don't use any type of barrel export and instead we generate here every single export

File: src/icons/index.ts [hunk 10]
```diff
@@ -0,0 +1,4 @@
+export { default as AngleRight } from "./AngleRight";
```
Comment (academo) [inline src/icons/index.ts:1]: let's move this file content to the one on src/index. There's no real need for this file.

**Review by academo — COMMENTED** [review]
Generally a good idea!

Things that would be good to have to make sure the project stays well maintained: 

* A linting script that makes sure every single SVG file has an icon component counterpart. This will prevent us from having abandoned SVG files we can't be sure if are used or not.
* A github action that runs linting and build scripts to make sure there's nothing broken
* Make 100% sure all this library is tree-shakeable and build integration tests for it. (e.g. a test webpack configuration with an app that uses only a few icons and the test makes sure the final output doesn't go beyond some limited amount of KBs)

File: src/icons/index.ts (hunk 10 above)
Comment (Clarity-89) [reply src/icons/index.ts:1]: This is a generated file by svgr, will check if it's possible to tweak the settings to move it.

File: src/IconBase.tsx (hunk 8 above)
Comment (Clarity-89) [reply src/IconBase.tsx:53]: Good point!

File: src/IconBase.tsx (hunk 5 above)
Comment (Clarity-89) [reply src/IconBase.tsx:17]: This is the same API as we have for the current Icon component, and it doesn't look like we needed any custom sizes so far. Plus, as an edge case, those can be tweaked via `className`.

**Review by joshhunt — COMMENTED** [review]
Looks good, nothing too objectionable here. Agree with the points from Esteban - we need more linting/ci/cd to make sure that any generated files are always 100% correct and in sync.

I also think that all generated files should be _very_ clearly marked so - either in a folder called `generated` or named `.gen.tsx`.

I still believe that the whole development flow would be easier if this package was in the Grafana repo. I see no advantages to having this external, in another repo, and only pains due to the hassle of yarn link etc.

**Comment by academo** [comment]
> I still believe that the whole development flow would be easier if this package was in the Grafana repo. I see no advantages to having this external, in another repo, and only pains due to the hassle of yarn link etc.

+1 on this. We already have all the SVG icon files in the grafana repo. We can move this there.

**Comment by Clarity-89** [comment]
> > I still believe that the whole development flow would be easier if this package was in the Grafana repo. I see no advantages to having this external, in another repo, and only pains due to the hassle of yarn link etc.
> 
> +1 on this. We already have all the SVG icon files in the grafana repo. We can move this there.

I'm working on moving this PR to grafana, making a separate package there. I don't think we'd reuse the actual svgs from Grafana though as it'll make the build process complicated, plus we want to restructure the svgs a bit and update names for some of them (e.g. remove `gf-` prefix for custom icons).

**Comment by Clarity-89** [comment]
Closing this in favor of https://github.com/grafana/grafana/pull/82314.

**Pull request closed without merging**